In [47]:
import numpy as np
import pandas as pd
from scipy import stats
from sklearn.datasets import load_iris

##### Exercise 2.

In [3]:
table = pd.read_html('https://en.wikipedia.org/wiki/Chi-squared_test')

In [4]:
table

[     Unnamed: 0    A    B    C    D  total
 0  White collar   90   60  104   95    349
 1   Blue collar   30   50   51   20    151
 2     No collar   30   40   45   35    150
 3         Total  150  150  200  150    650,
    .mw-parser-output .navbar{display:inline;font-size:88%;font-weight:normal}.mw-parser-output .navbar-collapse{float:left;text-align:left}.mw-parser-output .navbar-boxtext{word-spacing:0}.mw-parser-output .navbar ul{display:inline-block;white-space:nowrap;line-height:inherit}.mw-parser-output .navbar-brackets::before{margin-right:-0.125em;content:"[ "}.mw-parser-output .navbar-brackets::after{margin-left:-0.125em;content:" ]"}.mw-parser-output .navbar li{word-spacing:-0.125em}.mw-parser-output .navbar-mini abbr{font-variant:small-caps;border-bottom:none;text-decoration:none;cursor:inherit}.mw-parser-output .navbar-ct-full{font-size:114%;margin:0 7em}.mw-parser-output .navbar-ct-mini{font-size:114%;margin:0 4em}.mw-parser-output .infobox .navbar{font-size:100%}.mw-par

In [5]:
type(table)


list

In [6]:
df = pd.DataFrame(table[0])

In [7]:
df

Unnamed: 0    A    B    C    D  total
0  White collar   90   60  104   95    349
1   Blue collar   30   50   51   20    151
2     No collar   30   40   45   35    150
3         Total  150  150  200  150    650

In [8]:
names = []

values = []
for i in range(len(df)):
    names.append(df.iloc[i][0])
    values.append(np.array(df.iloc[i][1:-1]))
print(names, values)
combined = list(zip(names,values))
print('Combined',combined)


['White collar', 'Blue collar', 'No collar', 'Total'] [array([90, 60, 104, 95], dtype=object), array([30, 50, 51, 20], dtype=object), array([30, 40, 45, 35], dtype=object), array([150, 150, 200, 150], dtype=object)]
Combined [('White collar', array([90, 60, 104, 95], dtype=object)), ('Blue collar', array([30, 50, 51, 20], dtype=object)), ('No collar', array([30, 40, 45, 35], dtype=object)), ('Total', array([150, 150, 200, 150], dtype=object))]


In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  4 non-null      object
 1   A           4 non-null      int64 
 2   B           4 non-null      int64 
 3   C           4 non-null      int64 
 4   D           4 non-null      int64 
 5   total       4 non-null      int64 
dtypes: int64(5), object(1)
memory usage: 320.0+ bytes


In [10]:
print(combined[1][1])
print(values)

[30 50 51 20]
[array([90, 60, 104, 95], dtype=object), array([30, 50, 51, 20], dtype=object), array([30, 40, 45, 35], dtype=object), array([150, 150, 200, 150], dtype=object)]


In [11]:
stats.chi2_contingency(values)

(24.571202858582602,
 0.003483988784534318,
 9,
 array([[ 80.53846154,  80.53846154, 107.38461538,  80.53846154],
        [ 34.84615385,  34.84615385,  46.46153846,  34.84615385],
        [ 34.61538462,  34.61538462,  46.15384615,  34.61538462],
        [150.        , 150.        , 200.        , 150.        ]]))

### Chi Stat and Chi-square Test
**The Chi Stat is 24.57** from the above output which verifies the value given in the Wikipedia article. The Chi stat is calculated by analysing the difference between the observed and expected distributions.  Χ2 = Σ (O − E)2E / E

The chi-square test is used to test for a possible relationship between variables.  The more the observations differ from the expected distribution, the greater the chance that the variables are dependent.





### Degrees of Freedom
The degrees of freedom is 9 in this the above table. This is a measure of how many independent observations can be made. Generally, the higher the degrees of freedom is, the more relevant the Chi Square test will be.

Degrees of freedom is used in the calculation of the P value

### The P Value
**The P-value is 0.003** which indicates that neighbourhood (A,B,C,D) and type of worker living there are not independent. The difference between the expected distribution of types of worker in each area and the actual distribution is large enough to indicate that the neighbourhood and category of worker are

### References:

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html

https://towardsdatascience.com/running-chi-square-tests-in-python-with-die-roll-data-b9903817c51b

https://www.investopedia.com/terms/d/degrees-of-freedom.asp

https://towardsdatascience.com/chi-squared-test-for-feature-selection-with-implementation-in-python-65b4ae7696db

https://www.mathsisfun.com/data/chi-square-test.html

https://machinelearningmastery.com/chi-squared-test-for-machine-learning/



##### Exercise 3.

### Excel Function STDEV.P

The Excel function **STDEV.P** calculates the Standard Deviation when the values for every member of a population is known and we wish to know the standard deviation.  The function takes every member of an array representing the entire population as its arguments.  Importantly, in contrast with STDEV.S, this function uses the "n" method and the following formula √[∑(x - x̃)²/n]  .


### Excel Function STDEV.S

**STDEV.S** is an Excel function which calculates the Standard Deviation of a sample contained within a population. For most real-life scenarios, this is the best way to calculate the standard deviation as very often we do not have access to the entire target population.  This function uses the "n-1" method. "n-1" is a type of correction when calculating the standard deviation from only a sample of the total population.  The observed values are usually closer to the sample mean than to the real population mean.  Therefore using "n-1" corrects for this anticipated error by making the standard deviation larger.

### numpy Simulation



In [44]:
# Generate random 1d array as population


STDEV_P = 0
STDEV_S = 0
actual = 0
# calculate std dev of population using "n" method (default in numpy)
# this figure represents the actual standard deviation!


for i in range(10):
    # Generate random 1d array as population
    population = np.random.randint(1,1000,80)

    n_method = np.std(population)
    actual += n_method

    # take a sample of the population

    sample = np.random.choice(population,30)


    # calculate std deviation of sample using  "n" method


    n_method_sample = np.std(sample)
    STDEV_P += n_method_sample


    # calculate std deviatian of sample using "n-1"

    n_1_method_sample = np.std(sample, ddof =1)
    STDEV_S += n_1_method_sample

    
print("Actual Standard Deviation: {}\nSample Standard Deviation STDEV_P: {}\nSample Standard Deviation STDEV_S: {}".format(actual,STDEV_P,STDEV_S))

     




Actual Standard Deviation: 2879.7606598396387
Sample Standard Deviation STDEV_P: 2774.158624091815
Sample Standard Deviation STDEV_S: 2821.583574377375


In [20]:
# Generate random 1d array as population

population = np.random.randint(1,1000,80)
population

array([ 16, 590, 862, 469, 720,  43, 419, 253, 859, 248, 484, 141, 785,
       771, 111, 810, 208, 538, 488, 142, 843, 681, 881, 815, 582, 210,
       155, 218, 483, 580, 118, 867, 238, 606, 155, 770, 321, 588, 995,
       705, 785, 804, 890, 531, 793, 576, 995, 220, 933, 696, 177, 551,
       221, 384, 434, 305,  29, 769, 580, 896, 306, 818, 315, 973, 849,
        49, 639, 892, 967, 931, 930, 810, 608, 861, 388, 174, 584, 880,
       840, 540])

In [21]:
# calculate std dev of population using "n" method (default in numpy)
# this figure represents the actual standard deviation!

n_method = np.std(population)
n_method 

289.56308655239536

In [22]:
# take a sample of the population

sample = np.random.choice(population,30)
sample

array([810, 859, 210, 892, 840, 221, 861, 892, 177, 608, 881, 849, 538,
       582, 720, 892, 804, 305, 785, 861, 933, 469, 861, 588, 804, 238,
       785,  43, 880, 177])

In [41]:
# calculate std deviation of sample using  "n" method

n_method_sample = np.std(sample)
n_method_sample

285.6917219661781

In [42]:
# calculate std deviatian of sample using "n-1"

n_1_method_sample = np.std(sample, ddof =1)
n_1_method_sample

290.5756949277737

### References:

https://support.microsoft.com/en-us/office/stdev-p-function-6e917c05-31a0-496f-ade7-4f4e7462f285

https://stats.stackexchange.com/questions/3931/intuitive-explanation-for-dividing-by-n-1-when-calculating-standard-deviation#:~:text=The%20standard%20deviation%20calculated%20with%20a%20divisor%20of,the%20population%20from%20which%20the%20sample%20was%20drawn.

https://honingds.com/blog/python-standard-deviation/#:~:text=numpy%20uses%20population%20standard%20deviation%20by%20default%2C%20which,using%20numpy.std%20for%20a%20sample%20of%20the%20data.

https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.random.choice.html

#### Exercise 4.

In [51]:
data = load_iris()
data

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

https://www.datasciencelearner.com/k-means-clustering-in-python-label-dataset/#:~:text=%20K%20Means%20Clustering%20in%20Python%20%3A%20Label,can%20also%20use%20your%20own%20dataset.%20More%20